v2/internal/testcommon/samples_tester.go

```diff
@@ -0,0 +1,230 @@
+/*
+Copyright (c) Microsoft Corporation.
+Licensed under the MIT license.
+*/
+
+package testcommon
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"reflect"
+	"strings"
+
+	"github.com/emirpasic/gods/maps/linkedhashmap"
+	. "github.com/onsi/gomega"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apimachinery/pkg/util/yaml"
+
+	resources "github.com/Azure/azure-service-operator/v2/api/resources/v1beta20200601"
+	"github.com/Azure/azure-service-operator/v2/internal/resolver"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime"
+)
+
+const RefsPackage = "refs"
+const samplesPath = "../../config/samples"
+
+// exclusions slice contains resources to exclude from test as these resources run into
+// "runtime.notRegisteredErr" error from the Scheme.
+var exclusions = [...]string{
+        "batchaccountpool", 
+        "snapshot",
+	"sqldatabasecontainerthroughputsetting", 
+	"sqldatabasethroughputsetting",
+	"storageaccountsqueueservice",
+	"webtest", 
+	// excluding dbformysql/user as is not an ARM resource
+	"user",
+}
+
+type SamplesTester struct {
+	tc            *KubePerTestContext
+	decoder       runtime.Decoder
+	UseRandomName bool
+}
+
+type SampleObject struct {
+	SamplesMap   map[string]*linkedhashmap.Map
+	childObjects map[string]genruntime.ARMMetaObject
+}
+
+func NewSampleObject() SampleObject {
+	return SampleObject{
+		SamplesMap:   make(map[string]*linkedhashmap.Map),
+		childObjects: make(map[string]genruntime.ARMMetaObject),
+	}
+}
+
+func NewSamplesTester(tc *KubePerTestContext, useRandomName bool) *SamplesTester {
+	return &SamplesTester{
+		tc:            tc,
+		decoder:       serializer.NewCodecFactory(tc.GetScheme()).UniversalDecoder(),
+		UseRandomName: useRandomName,
+	}
+}
+
+func (t SamplesTester) LoadSamples(rg *resources.ResourceGroup, group string) (*SampleObject, *SampleObject) {
+
+	samples := NewSampleObject()
+	depSamples := NewSampleObject()
+
+	groupPath := getGroupPath(group)
+
+	err := filepath.Walk(groupPath,
+		func(filePath string, info os.FileInfo, err error) error {
+
+			if err != nil {
+				return err
+			}
+			if !info.IsDir() && !isExclusion(filePath) {
+				//if !info.IsDir() && !isExclusion(filePath) {
+
+				sample := t.getObjectFromFile(filePath)
+
+				sample.SetNamespace(t.tc.Namespace)
+
+				if !strings.Contains(filePath, RefsPackage) {
+					if t.UseRandomName {
+						sample.SetName(t.tc.NoSpaceNamer.GenerateName(""))
+					}
+					handleObject(sample, rg, &samples)
+				} else {
+					// We don't change name for dependencies as they have fixed refs which we can't access inside the object
+					// need to make sure we have right names for the refs in samples
+					handleObject(sample, rg, &depSamples)
```

Comment:
Technically speaking I think you could access and change the refs, but we'd need to write a reflect helper to do it and it'd be a bit of a pain. This is OK for now but might come back to bite us if things aren't uniquely named.

Reply:
Yes, we could? The only thing I can think of is as each resource has uniquely named field for refs so for example: For EventSubscription it looks like `destination/storageQueueEventSubscriptionDestination/properties/resourceReference` and for VirtualMachine its `networkProfile/networkInterfaces/reference`. 

Which gives a hint that we'll have to iterate through each field using reflect and find out where we want to put the refs?

Reply:
Yeah I think this is something to maybe file an item for, not something to do right now